nifi-nar-bundles/nifi-email-bundle/nifi-email-processors/src/main/java/org/apache/nifi/processors/email/ListenSMTP.java

```diff
@@ -0,0 +1,383 @@
+        if (sslContextService != null) {
+            server.setEnableTLS(true);
+        } else {
+            server.setHideTLS(true);
+        }
+
+        // Set TLS to required in case CLIENT_AUTH = required
+        if (context.getProperty(CLIENT_AUTH).getValue() == "REQUIRED") {
+            server.setRequireTLS(true);
+        }
+
+        this.server = server;
+        server.start();
+
+        getLogger().info("Server started and listening on port " + server.getPort());
+
+        initialized.set(true);
+    }
+
+    @OnStopped
+    public void shutdown() throws Exception {
+        if (server != null) {
+            getLogger().debug("Shutting down server");
+            server.stop();
+            getLogger().info("Shut down {}", new Object[]{server});
+        }
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        try {
+            initializeSMTPServer(context);
+        } catch (Exception e) {
+            context.yield();
+            throw new ProcessException("Failed to initialize the SMTP server", e);
+        }
+
+        if (messages.isEmpty()) {
+            return;
+        }
+
+        FlowFile flowfile = session.create();
+        SmtpEvent message;
+
+        try {
+            message = messages.take();
+
+            if (message.getBodyData() != null) {
+                byte[] messageData = message.getBodyData();
+                flowfile = session.write(flowfile, new OutputStreamCallback() {
+
+                    // Write the messageData to flowfile content
+                    @Override
+                    public void process(OutputStream out) throws IOException {
+                        out.write(messageData);
+                    }
+                });
+            }
+
+            HashMap<String, String> attributes = new HashMap<>();
+            // Gather message attributes
+            attributes.put(SMTP_HELO, message.getHelo());
+            attributes.put(SMTP_FROM, message.getFrom());
+            attributes.put(SMTP_TO, message.getTo());
+
+            List<Map> details = message.getCertifcateDetails();
+            int c = 0;
+
+            // Add a selection of each X509 certificates to the already gathered attributes
+
+            for (Map<String, String> detail : details) {
+                attributes.put("smtp.certificate." + "c".toString() + "serial", detail.getOrDefault("SerialNumber", null));
+                attributes.put("smtp.certificate." + "c".toString() + "subjectName", detail.getOrDefault("SubjectName", null));
+                c++;
+            }
+
+            // Set Mime-Type
+            attributes.put(CoreAttributes.MIME_TYPE.key(), MIME_TYPE);
+
+            // Add the attributes. to flowfile
+            flowfile = session.putAllAttributes(flowfile, attributes);
+            session.getProvenanceReporter().receive(flowfile, "smtp://" + message.getRemoteIP() + "/");
```

Comment:
When I tested locally, the transit URL was set as `smtp:///127.0.0.1:61579/`, that was sender's address. I checked other components like Site-to-Site and ListenHttp, usually transit URL represents "what component received the incoming flow file" for `receive` events, rather than "where it came through". Using SMTP_HOSTNAME and SMTP_PORT to construct transit URL will make it align with other components and provide more useful information to users.